interface.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package pubsub
+
+// Topic represents a message that can be subscribed to.
+type Topic string
+
+// TopicMatcher defines the Match method that is used to determine
+// if the subscriber should be notified about a particular message.
+type TopicMatcher interface {
+	Match(Topic) bool
+}
+
+// Hub represents an in-process delivery mechanism. The hub maintains a
+// list of topic subscribers.
+type Hub interface {
+
+	// Publish will notifiy all the subscribers that are interested by calling
+	// their handler function.
+	Publish(topic Topic, data interface{}) (Completer, error)
+
+	// Subscribe takes a topic matcher, and a handler function. If the matcher
+	// matches the published topic, the handler function is called. If the
+	// handler function does not match what the Hub expects an error is
+	// returned. The definition of the handler function depends on the hub
+	// implementation. Please see NewSimpleHub and NewStructuredHub.
+	Subscribe(matcher TopicMatcher, handler interface{}) (Unsubscriber, error)
```

Comment:
I'm not keen at all on the handler reflection.

Do you really need one interface to represent two types of hubs? Is there even a need for an interface in this package at all? Why not just expose SimpleHub and StructuredHub?